STACpopulator/extensions/base.py

```diff
@@ -0,0 +1,331 @@
+    extensions : list[str]
+        Name of the class attributes that point to STAC extension helper classes. Those extension classes should have an `apply` method.
+    """
+    # STAC item properties
+    geometry: AnyGeometry | None
+    bbox: list[float]
+    start_datetime: datetime
+    end_datetime: datetime
+
+    model_config = ConfigDict(populate_by_name=True, extra="ignore", arbitrary_types_allowed=True)
+
+    # Helpers are automatically detected by being Helper subclasses
+    _helpers: list[str] = PrivateAttr([])
+
+    @property
+    def uid(self) -> str:
+        """Return a unique ID. When subclassing, use a combination of properties uniquely identifying a dataset."""
+        # TODO: Should this be an abstract method?
+        import uuid
+        return str(uuid.uuid4())
+
+    @model_validator(mode="after")
+    def find_helpers(self):
+        """Populate the list of extensions."""
+        for key, field in self.model_fields.items():
+            if isinstance(field.annotation, type) and issubclass(field.annotation, Helper):
+                self._helpers.append(key)
+
+    def stac_item(self) -> "pystac.Item":
+        """Create a STAC item and add extensions."""
+        item = pystac.Item(
+            id=self.uid,
+            geometry=self.geometry.model_dump(),
+            bbox=self.bbox,
+            start_datetime=self.start_datetime,
+            end_datetime=self.end_datetime,
+            datetime=None,
+            properties={},
+        )
+
+        # Add extensions
+        for ext in self._helpers:
+            getattr(self, ext).apply(item)
+
+        try:
+            item.validate()
+        except STACValidationError as e:
+            raise Exception("Failed to validate STAC item") from e
+
+        return json.loads(json.dumps(item.to_dict()))
+
+
+def metacls_extension(name, schema_uri):
+    """Create an extension class dynamically from the properties."""
+    cls_name = f"{name.upper()}Extension"
+
+    bases = (MetaExtension,
+             Generic[T],
+             PropertiesExtension,
+             ExtensionManagementMixin[Union[pystac.Asset, pystac.Item, pystac.Collection]]
+             )
+
+    attrs = {"name": name, "schema_uri": schema_uri}
+    return types.new_class(name=cls_name, bases=bases, kwds=None, exec_body=lambda ns: ns.update(attrs))
+
+
+class MetaExtension:
+    name: str
+    schema_uri: str
+
+    def apply(self, properties: dict[str, Any]) -> None:
+        """Applies CMIP6 Extension properties to the extended
+        :class:`~pystac.Item` or :class:`~pystac.Asset`.
+        """
+        for prop, val in properties.items():
+            self._set_property(prop, val)
+
+    @classmethod
+    def get_schema_uri(cls) -> str:
+        """We have already validated the JSON schema."""
+        return cls.schema_uri
+
+    @classmethod
+    def has_extension(cls, obj: S):
+        # FIXME: this override should be removed once an official and versioned schema is released
```

Comment:
Does this comment apply here? Is this copied from the CMIP6 code?

Do we need to keep this function since the only use of it seems to be commented out in the CMIP6 code anyway.
And if we can get rid of it can we also get rid of `get_schema_uri`?